infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/Bootstrapper.java

```diff
@@ -0,0 +1,176 @@
+    private final    String                                   wsIntallerEndpoint;
+    private final    CountDownLatch                           bootstraplatch;
+    private final    AtomicInteger                            endpointId;
+    private final    EventSubscriber<BootstrapperStatusEvent> bootstrapperStatusListener;
+    private volatile BootstrapperStatusEvent                  resultEvent;
+
+    @Inject
+    public Bootstrapper(@Assisted String machineName,
+                        @Assisted RuntimeIdentity runtimeIdentity,
+                        @Assisted DockerMachine dockerMachine,
+                        EventService eventService,
+                        @Named("che.workspace.che_server_endpoint") String apiEndpoint) {
+        this.eventService = eventService;
+        this.machineName = machineName;
+        this.runtimeIdentity = runtimeIdentity;
+        this.dockerMachine = dockerMachine;
+        this.bootstraplatch = new CountDownLatch(1);
+        this.endpointId = new AtomicInteger();
+        this.wsIntallerEndpoint = apiEndpoint.replace("http", "ws").replace("wsmaster/api", "wsmaster")
+                                  + "/installer/websocket/";
+        this.bootstrapperStatusListener = event -> {
+            BootstrapperStatus status = event.getStatus();
+
+            //skip starting status event
+            if (status.equals(BootstrapperStatus.DONE) || status.equals(BootstrapperStatus.FAILED)) {
+                //check boostrapper belongs to current runtime and machine
+                RuntimeIdentityDto runtimeId = event.getRuntimeId();
+                if (event.getMachineName().equals(machineName)
+                    && runtimeIdentity.getEnvName().equals(runtimeId.getEnvName())
+                    && runtimeIdentity.getOwner().equals(runtimeId.getOwner())
+                    && runtimeIdentity.getWorkspaceId().equals(runtimeId.getWorkspaceId())) {
+
+                    resultEvent = event;
+                    bootstraplatch.countDown();
+                }
+            }
+        };
+    }
+
+    /**
+     * Bootstraps agents installers and wait while they finished.
+     *
+     * @param agents
+     *         agents to install
+     * @throws InfrastructureException
+     *         when bootstrapping timeout reached
+     * @throws InfrastructureException
+     *         when bootstrapping failed
+     * @throws InfrastructureException
+     *         when any other error occurs while bootstrapping
+     */
+    public void bootstrap(List<AgentImpl> agents) throws InfrastructureException {
+        injectBootstrapper(agents);
+
+        this.eventService.subscribe(bootstrapperStatusListener, BootstrapperStatusEvent.class);
+        boolean bootstrappingFinished;
+        try {
+            String endpoint = wsIntallerEndpoint + endpointId.getAndIncrement();
+            dockerMachine.exec(BOOTSTRAPPER_BASE_DIR + BOOTSTRAPPER_FILE +
+                               " -machine-name " + machineName +
```

Comment:
No, i meant pretty the same thing, i think it can be less verbose e.g.
```java
dockerMachine.exec(BootstrapperCmd.binary(BOOTSTRAPPER_DIR + BOOTSTRAPPER_FILE)
                                  .file(BOOTSTRAPPER_DIR + CONFIG_FILE)
                                  .machine(machineName)
                                  .runtime(runtimeIdentity)
                                  .pushEndpoint(endpoint)
                                  .pushLogsEndpoint(endpoint)
                                  .get(), null);

```

Where bootstrapper cmd can be something like [this](https://gist.github.com/evoevodin/edec8ba1172c474b753b2f996b284284).

Up to you whether to change it or not.